### PR TITLE
Feature/access reel named path

### PR DESCRIPTION
**Pull request by goto40**
See #303 

Here the find functions in the rrel module are enhanced to allow to access the path (see, e.g.,  `test_referencing_attributes_with_rrel_and_full_path_access()`).

**Use Case**
 * I have a model where I model structures with attributes, like https://github.com/textX/textX/blob/master/tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py
 * I use the model to generate code, where I access referenced elements: e.g. from the `Reference` rule: `reference d.c.b.a.x`. For this I need the full path for code generation. But this path is not accessible via RREL (unlike the custom scoping using complicated rules (`RefItem`, https://github.com/textX/textX/blob/master/tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py#L76) to model the path).

**Proposed extension:**
A new flag 'p' to enable the use of a reference proxy. This proxy allows to access the path of named objects forming the reel resolution. 

## Code review checklist

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [x] Tests have been included and/or updated
- [x] Docstrings have been included and/or updated, as appropriate
- [x] Standalone docs have been updated accordingly
- [x] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`, no need
      to update for typo fixes and such).

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by goto40** [comment]
I picked up your suggestion and added a wrapper (`ReferenceProxy`) around it to support our type semantics.

@igordejanovic 
Maybe you can have a look, if the proxy could be done better (maybe we need to add more methods, like `__eq__`?)

**Comment by goto40** [comment]
If we converge to a solution, I will add more unittests...

File: tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py [hunk 1]
```diff
@@ -268,3 +269,62 @@ def test_referencing_attributes_with_rrel_all_in_one_splitstring():
...
+        Val:
+            'val' name=ID (':' type=[Struct])?;
+        Instance:
+            'instance' name=ID (':' type=[Struct])?;
+        Reference:
+            'reference' ref=[Val|FQN|+p:instances.~type.vals.(~type.vals)*];
+        FQN: ID ('.' ID)*;
+        ''')
+    m = mm.model_from_str(model_text)
+    m.references[-1].ref == m.structs[0].vals[0]  # a.x
```
Comment (igordejanovic) [inline tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py]: This is already tested in line 307 bellow.

File: textx/scoping/rrel.py [hunk 2]
```diff
@@ -56,7 +56,8 @@ class RRELBase(object):
     def __init__(self):
         pass
 
-    def get_next_matches(self, obj, lookup_list, allowed, first_element=False):
+    def get_next_matches(self, obj, lookup_list, allowed, matched_path_so_far,
```
Comment (igordejanovic) [inline textx/scoping/rrel.py]: I propose shortening `matched_path_so_far` -> `matched_path` in all places. It is easier to read/write.

**Review by igordejanovic — COMMENTED** [review]
@goto40  Very good!  See inline comments for some minor issues, but overall good to go.

File: tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py (hunk 1 above)
Comment (goto40) [reply tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py]: (deleted)

File: textx/scoping/rrel.py (hunk 2 above)
Comment (goto40) [reply textx/scoping/rrel.py]: (done)

File: docs/rrel.md [hunk 3]
```diff
@@ -153,6 +153,71 @@ When using this extra prefix the importURI feature is activated
...
+    'val' name=ID (':' type=[Struct])?;
+Instance:
+    'instance' name=ID (':' type=[Struct])?;
+Reference:
+    'reference' ref=[Val|FQN|+p:instances.~type.vals.(~type.vals)*];
+FQN: ID ('.' ID)*;
+```
+
+The **use case** for that feature is that you sometimes need
+to access all model elements specfied in a model reference. Consider
```
Comment (igordejanovic) [inline docs/rrel.md]:
```suggestion
to access all model elements specified in a model reference. Consider
```

File: tests/functional/test_scoping/test_rrel.py [hunk 4]
```diff
@@ -269,8 +279,82 @@ def test_rrel_repetitions():
...
+            }
+            A a2 {
+                A aa2 {}
+            }
+            R a2::unknown
+        ''')
+
+
+def test_split_str_multifile():
+    # same as test abive, but with "+mp:" flag
```
Comment (igordejanovic) [inline tests/functional/test_scoping/test_rrel.py]:
```suggestion
    # same as test above, but with "+mp:" flag
```

**Review by igordejanovic — APPROVED** [review]
Overall looks very good. I caught two typos (see bellow).
Fell free to squash merge or rework commits however you see fit.

File: docs/rrel.md (hunk 3 above)
Comment (goto40) [reply docs/rrel.md]: done

File: tests/functional/test_scoping/test_rrel.py (hunk 4 above)
Comment (goto40) [reply tests/functional/test_scoping/test_rrel.py]: done

**Pull request merged**